File: 김지원/스터디자료/스터디프로그래머스/5주차_김지원/5-2/5-2.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <string.h>

#define SIZE 10001

typedef struct
{
	int queue[SIZE];
	int front, rear;
}QueueType;

void Queue_init(QueueType* q)
{
	q->front = q->rear = 0;
}

int is_empty(QueueType* q)
{
	return (q->front == q->rear);
}

int is_full(QueueType* q)
{
	return (q->front == (q->rear + 1) % SIZE);
}

void enqueue(QueueType* q, int e)
{
	if (is_full(q))
		return;

	q->rear = (q->rear + 1) % SIZE;
	q->queue[q->rear] = e;
}

int dequeue(QueueType* q)
{
	if (is_empty(q))
		return -1;

	q->front = (q->front + 1) % SIZE;
	return q->queue[q->front];
}

int front(QueueType* q)
{
	if (is_empty(q))
		return -1;
	return q->queue[q->front + 1];
}

int back(QueueType* q)
{
	if (is_empty(q))
		return -1;
	return q->queue[q->rear];
}

int size(QueueType* q)
{
	if (q->front < q->rear)
		return q->rear - q->front;
	else
		return SIZE - q->front + q->rear;
}

int main()
{
	QueueType queue;
	Queue_init(&queue);

	int i, n, e;
	char op[6];

	scanf("%d", &n);

	for (i = 0; i < n; i++)
	{
		scanf(" %s", op);

		if (strcmp(op, "push") == 0)
		{
			scanf(" %d", &e);
			enqueue(&queue, e);
		}
		else if (strcmp(op, "pop") == 0)
		{
			e = dequeue(&queue);
			printf("%d\n", e);
		}
		else if (strcmp(op, "front") == 0)
		{
			e = front(&queue);
			printf("%d\n", e);
		}
		else if (strcmp(op, "back") == 0)
		{
			e = back(&queue);
			printf("%d\n", e);
		}
		else if (strcmp(op, "empty") == 0)
		{
			e = is_empty(&queue);
			printf("%d\n", e);
		}
		else if (strcmp(op, "size") == 0)
		{
			e = size(&queue);
			printf("%d\n", e);
		}
	}

	return 0;
}

File: 김지원/스터디자료/스터디프로그래머스/5주차_김지원/5-2/5-2.js
const readline = require("readline");

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
});

let input = [];

rl.on("line", function (line) {
	input.push(line);
}).on("close", function () {
	//console.log(input);
	solve();
	process.exit();
});
class NodeQueue {
	constructor(value) {
		this.next = null;
		this.value = value;
	}
}
class Queue {
	constructor() {
		this.head = null;
		this.tail = null;
		this.size = 0;
	}
	enqueue(value) {
		let nodeQueue = new NodeQueue(value);
		if (this.size == 0) {
			this.head = nodeQueue;
		} else {
			this.tail.next = nodeQueue;
		}
		this.tail = nodeQueue;
		this.size++;
	}
	dequeue() {
		if (this.size == 0) {
			return null;
		}
		let value = this.head.value;
		this.head = this.head.next;
		this.size--;
		if (this.size == 0) {
			this.tail = null;
		}
		return value;
	}
	isEmpty() {
		return this.size == 0;
	}
	front() {
		if (this.size == 0) {
			return null;
		}
		return this.head.value;
	}
	back() {
		if (this.size == 0) {
			return null;
		}
		return this.tail.value;
	}
	getSize() {
		return this.size;
	}
}

//Array로 구현 pop의 시간복잡도가 O(n)임
class Queue2 {
	constructor() {
		this.q = [];
	}
	enqueue(value) {
		this.q.push(value);
	}
	dequeue() {
		if (this.q.length == 0) {
			return null;
		}
		return this.q.shift();
	}
	isEmpty() {
		return this.q.length == 0;
	}
	front() {
		if (this.q.length == 0) {
			return null;
		}
		return this.q[0];
	}
	back() {
		if (this.q.length == 0) {
			return null;
		}
		return this.q[this.q.length - 1];
	}
	getSize() {
		return this.q.length;
	}
}

function solve() {
	let output = "";
	let q = new Queue();
	for (let i = 1; i < input.length; i++) {
		let line = input[i];
		if (line.startsWith("push")) {
			let spt = line.split(" ");
			q.enqueue(spt[1]);
		} else if (line.startsWith("pop")) {
			if (q.isEmpty()) {
				output += -1 + "\n";
			} else {
				output += q.dequeue() + "\n";
			}
		} else if (line.startsWith("front")) {
			if (q.isEmpty()) {
				output += -1 + "\n";
			} else {
				output += q.front() + "\n";
			}
		} else if (line.startsWith("size")) {
			output += q.getSize() + "\n";
		} else if (line.startsWith("empty")) {
			if (q.isEmpty()) output += 1 + "\n";
			else {
				output += 0 + "\n";
			}
		} else if (line.startsWith("back")) {
			if (q.isEmpty()) {
				output += -1 + "\n";
			} else {
				output += q.back() + "\n";
			}
		}
	}
	console.log(output.trimEnd());
}
